pkg/ebpf/runtime_asset.go

```diff
@@ -0,0 +1,90 @@
+// +build linux_bpf
+
+package ebpf
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/DataDog/datadog-agent/pkg/ebpf/compiler"
+	"github.com/DataDog/datadog-agent/pkg/util/kernel"
+)
+
+type CompiledOutput interface {
+	io.Reader
+	io.ReaderAt
+	io.Closer
+}
+
+// RuntimeAsset represents an asset that needs it content integrity checked at runtime
+type RuntimeAsset struct {
+	filename string
+	hash     string
+}
+
+func NewRuntimeAsset(filename, hash string) *RuntimeAsset {
+	return &RuntimeAsset{
+		filename: filename,
+		hash:     hash,
+	}
+}
+
+// Verify reads the asset in the provided directory and verifies the content hash matches what is expected.
+// On success, it returns the full path and content hash of the asset.
+func (a *RuntimeAsset) Verify(dir string) (string, string, error) {
+	p := filepath.Join(dir, "runtime", a.filename)
+	f, err := os.Open(p)
+	if err != nil {
+		return "", "", err
+	}
+	defer f.Close()
+
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return "", "", fmt.Errorf("error hashing file %s: %w", f.Name(), err)
+	}
+	if fmt.Sprintf("%x", h.Sum(nil)) != a.hash {
+		return "", "", fmt.Errorf("file content hash does not match expected value")
+	}
+	return p, a.hash, nil
+}
+
+// Compile compiles the runtime asset if necessary and returns the resulting file.
+func (a *RuntimeAsset) Compile(config *Config, cflags []string) (CompiledOutput, error) {
+	kv, err := kernel.HostVersion()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get kernel version: %w", err)
+	}
+
+	inputFile, hash, err := a.Verify(config.BPFDir)
```

Comment:
I think that there is a [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) issue. First, the hash is checked, then the file is opened again and compiled. It leaves an opportunity to modify the file in between these two operations.

Reply:
You are correct. I will need to modify the compiler to support in-memory buffers rather than files, but we have talked about wanting that ability anyways.

Reply:
Updated to use in-memory buffers. This should address the TOCTOU issue.